routine_butler/main.py

```diff
@@ -25,67 +35,77 @@ def set_colors():
-        set_colors()
+    async def __call__(self, scope, receive, send):
+        scope['db_engine'] = self.engine
+        await self.app(scope, receive, send)
 
-        with ui.card():
-            ui.label("Login")
-            ui.separator()
-            username_input = ui.input("Username")
-            login_button = ui.button("Login")
+def db_session(request: Request) -> Session:
+    """helper to get DB session from request"""
+    return Session(request.scope['db_engine'])
 
-        login_button.on(
-            "click", lambda: self.on_login_attempt(username_input.value)
-        )
 
-    def main_gui(self):
-        assert self.user is not None
+@ui.page('/')
+def main_gui(request: Request) -> None:
+    set_colors()
 
-        set_colors()
+    if not is_authenticated(request):
+        return RedirectResponse('/login')
 
-        header = Header()
-        routines_sidebar = RoutinesSidebar(
-            user=self.user, repository=self.repository
-        )
-        programs_sidebar = ProgramsSidebar(
-            user=self.user, repository=self.repository
-        )
+    with db_session(request) as session:
+        username = session_info[request.session['id']]['username']
+        user = User.eagerly_get_user(session, username)
 
+        if user is None:  # invalid session-id
+            del request.session['id']
+            ui.open('/login')
+
+        header = Header()
+        routines_sidebar = RoutinesSidebar(session, user=user)
         header.routines_button.on("click", routines_sidebar.toggle)
-        header.programs_button.on("click", programs_sidebar.toggle)
 
-    def on_login_attempt(self, username):
-        with self.repository.session() as session:
-            user = self.repository.eagerly_get_user(username, session=session)
+        # programs_sidebar = ProgramsSidebar(
+        #     user=self.user, repository=self.repository
+        # )
+
+        # header.programs_button.on("click", programs_sidebar.toggle)
+
+
+@ui.page('/login')
+def login(request: Request) -> None:
+    set_colors()
+    def on_login_attempt():
+        with db_session(request) as session:
+            user = User.eagerly_get_user(session, username_input.value)
 
         if user is None:
             ui.notify("Invalid username")
         else:
-            self.user = user
-            self.repository.current_username = user.username
-            self.main_frame.clear()
-            with self.main_frame:
-                self.main_gui()
+            session_info[request.session['id']] = {'username': user.username, 'authenticated': True}
             ui.notify("Welcome, " + user.username + "!")
-
-
-def main(repository: Repository, user: Optional[User] = None):
-    RoutineButler(repository, user)
+            ui.open('/')
+
+    # used with --testing for automatically login users
+    if request.app.auto_login is not None:
+        session_info[request.session['id']] = {'username': request.app.auto_login, 'authenticated': True}
```

Comment:
@schettino72, when I run `python3 run.py --testing`, I get `KeyError: 'id'` at this line. At this point, `request.session` is `{}` (an empty dict)